### ISSUE TITLE
Nice function to get the Monad value

### ISSUE THREAD

**Issue by afharo** [gcanti/fp-ts#225]
I have this use case:
```typescript
import {Option, some, none} from "fp-ts/lib/Option";
interface TestObj {
    a: number,
    b: Option<number>
}

const pairs: Array<TestObj> = [
  {a: 1, b: some(1)}, 
  {a: 1, b: none}, 
];

function add(a: number, b: number) {
   return a+b;
}

// I want to add only the ones when b is some, just discard the rest
pairs
   .filter((p) => p.b.isSome())
   .map((p) => add(p.a, p.b.getOrElse(() => 0));
```
`getOrElse` function will never be hit, but I need it to be there if I want to get the value from b out of the Option. I know it exists b._A but the underscore kind of tells it's an internal value and shouldn't be used, doesn't it?
Is there any other way to solve this or can it be added a public function/property to get the value?

**Comment by gcanti** [comment]
> I know it exists b._A

`_A` is a "phantom type", i.e. a field that doesn't contain any value (it's there for type checking purposes, also it helps to implement HKTs)

You can use `Option#fold` I guess

```ts
pairs.reduce((acc, obj) => acc + obj.b.fold(() => 0, n => n + obj.a), 0)
```

**Comment by raveclassic** [comment]
or even shorter:
```
pairs.reduce((acc, obj) => acc + obj.b.isSome() ? obj.a : 0, 0)
```

**Comment by raveclassic** [comment]
@afharo I think it's the only way because `filter` does not tell type-system that all filtered values are now `Some<number>`s, they are still `Option<number>`. So if you you need to add `a` to `b` you still need to check wether `b` is a `Just` when extracting its actual value. You can do it only with `getOrElse`.

**Comment by afharo** [comment]
Yeah, I can use those, but, at the end either `.fold` or `.getOrElse` is all about providing a value when I already know that value is enforced to exist, making the code cumbersome and providing a branch that will never be covered.
The adding example was a very simple one to illustrate the example, but there are some other use cases like (if all should have a some() element in b)

```typescript
// If at least 1 is none
if (pairs.find((p) => p.b.isNone()) {
   return doSomething()
} else {
  // All values in here are some, fold or getOrElse are redundant.
  return doSomethingUsingB()
}
```

Another use case where `fold` or `getOrElse` are not the best option:
```typescript
const pairs: Array<TestObj> = [
  {a: 1, b: some({c: "test", d:3})}, 
  {a: 1, b: none}, 
];

// I need to apply casting
pairs
   .filter((p) => p.b.isSome())
   .map((p) => doWhatEver(p.a, p.b.getOrElse(() => ({} as PairsBInterface))); 

// Or build a similar object
pairs
   .filter((p) => p.b.isSome())
   .map((p) => doWhatEver(p.a, p.b.getOrElse(() => ({c:"", d:0}))); 
```
Any of those options: apply casting or build a similar object don't seem to me like a valid option because none of them are valid. Yeah, I know it won't ever get in there but when reading the code, it kind of tells it could happen and, if it does happen, it will pass either a non-type-matching var (in running time) or a randomly invented value.

**Comment by afharo** [comment]
@raveclassic 
> I think it's the only way because filter does not tell type-system that all filtered values are now Some<number>s, they are still Option<number>. 

Could this be because (from https://github.com/gcanti/fp-ts/blob/master/src/Option.ts#L199)
```typescript
  isSome(): boolean {
    return true
  }
```
Shouldn't it be something like:
```typescript
  isNone(): this is None<A> {
    return true
  }
  isSome(): this is Some<A> {
    return true
  }
```
From my tests, it still fails because of Array.filter typing... but it works with simple ifs ;)

**Comment by gcanti** [comment]
FWIW there's a top level [isSome](https://github.com/gcanti/fp-ts/blob/4f8cf76809d92a9658d682267c655b75d1dd49d3/src/Option.ts#L263) which is a type predicate and AFAIK `filter` has an overloading which understands type predicates

```ts
const xs = [some(1), none]
// ys: Some<number>[]
const ys = xs.filter(isSome)
```

@afharo you are traversing `pairs` twice (`filter` + `map`). `reduce` + `fold` let you traverse only once in a type-safe way

**Comment by afharo** [comment]
@gcanti I did the reduce + fold option but then got some other typing inheritance issue (still to work on that one, though... ). Thank you for the suggestion.
I've noticed about the top level isSome and used it somewhere else (where the typing is properly inherited (not in filter)) 
Some sort of 
```typescript
if (isNone(bar)) {
   return endOfProcess()
}
...
a ton of logic...
...
callFunction(bar.value) // It works because the compiler understands bar is a Some at this point
```

**Comment by raveclassic** [comment]
@afharo You can call `toNullable` and then coerce to non-nullable type with `!`:
```typescript
const foo = fromNullable(123); //Option<number>
const fooExact = foo.toNullable()!; //number
```

**Comment by afharo** [comment]
@raveclassic, that's actually a good move! thanks :)

**Comment by gcanti** [comment]
@raveclassic `!` is awfully unsafe and is strictly forbidden in this repo 😛 

If you want to go fancy, you could use `filterMap` (though is traversing twice)

```ts
import { filterMap } from 'fp-ts/lib/Filterable'
import * as array from 'fp-ts/lib/Array'

filterMap(array)<TestObj, number>(obj => obj.b.map(n => n + obj.a))(pairs).reduce(add, 0)
```

**Comment by raveclassic** [comment]
@gcanti I understand it and I'd use `getOrElse` with `reduce` =)

**Comment by raveclassic** [comment]
@gcanti Maybe `toNullable` should be dropped at all, what was the point in adding it? It's unsafe and moves developers to bypass type constraints.

**Comment by vegansk** [comment]
@raveclassic , `toNullable` can be useful when working with the external APIs. Without it you will use `fold`. Unfortunately there is no function `toUndefined` that can be used for that. For example, when creating properties for react component. @gcanti , will you accept the PR with it's implementation?

**Comment by gcanti** [comment]
@raveclassic for interop
@vegansk yes

**Comment by raveclassic** [comment]
@vegansk, @gcanti I see. But isn't `getOrElse(constant(null))` or `getOrElse(constant(undefined))` enough as it's more general?

**Comment by raveclassic** [comment]
@gcanti Also why does `getOrElse` accept a lazy value? Was it an intentional decision?

**Comment by sledorze** [comment]
@raveclassic Immagine the alternate value being the result of a long computation, it would be nice to have it being executed *only* when required.

**Comment by raveclassic** [comment]
@sledorze Well, the typical usage is to provide a quick default value to get _wrapped_ value from `Option`. At least in my practise I always end up with `constant` call. But I got the point though, thx

**Comment by vegansk** [comment]
@raveclassic, your method doesn't work. The code:

```ts
let x: string | undefined = "a"

let y = option.fromNullable(x).getOrElse(constant(null))
```

fails to compile

```
ERROR in ./src/core/utils/fp.spec.ts
(80,46): error TS2345: Argument of type 'Lazy<null>' is not assignable to parameter of type 'Lazy<string>'.
  Type 'null' is not assignable to type 'string'
```

**Comment by raveclassic** [comment]
@vegansk Oops :) Now it does seem like a good place for `toNullable`/`toUndefined`
